x-pack/plugins/endpoint/server/routes/alerts/lib/index.ts

```diff
@@ -3,28 +3,97 @@
  * or more contributor license agreements. Licensed under the Elastic License;
  * you may not use this file except in compliance with the Elastic License.
  */
+import { RisonValue, decode, encode } from 'rison-node';
 import { SearchResponse } from 'elasticsearch';
-import { IScopedClusterClient } from 'kibana/server';
+import { stringify } from 'query-string';
+import { IScopedClusterClient, KibanaRequest } from 'kibana/server';
 import { JsonObject } from '../../../../../../../src/plugins/kibana_utils/public';
-import { esKuery, esQuery } from '../../../../../../../src/plugins/data/server';
+import { esKuery, esQuery, Filter, TimeRange } from '../../../../../../../src/plugins/data/server';
 import { AlertEvent, Direction, EndpointAppConstants } from '../../../../common/types';
+import { EndpointAppContext } from '../../../types';
 import {
-  AlertSearchQuery,
+  AlertListRequestQuery,
+  AlertSearchParams,
   AlertSearchRequest,
   AlertSearchRequestWrapper,
   AlertSort,
+  UndefinedResultPosition,
 } from '../types';
-
 export { Pagination } from './pagination';
 
-function reverseSortDirection(order: Direction): Direction {
-  if (order === Direction.asc) {
-    return Direction.desc;
+/**
+ * Translates API search, sort, and pagination parameters to ES-ready parameters.
+ */
+export const getAlertSearchParams = async (
+  request: KibanaRequest<unknown, AlertListRequestQuery, unknown>,
+  endpointAppContext: EndpointAppContext
+): Promise<AlertSearchParams> => {
+  const config = await endpointAppContext.config();
+  const reqData: AlertSearchParams = {
+    // Filtering
+    query: request.query.query,
+    filters:
+      request.query.filters !== undefined
+        ? ((decode(request.query.filters) as unknown) as Filter[])
+        : ([] as Filter[]),
+    dateRange: ((request.query.date_range !== undefined
+      ? decode(request.query.date_range)
+      : config.alertResultListDefaultDateRange) as unknown) as TimeRange,
+
+    // Sorting
+    sort: request.query.sort,
+    order: request.query.order,
+
+    // Paging - Simple
+    pageIndex: request.query.page_index,
+    pageSize: request.query.page_size,
+
+    // Paging - Cursor-Based
+    searchAfter: request.query.after,
+    searchBefore: request.query.before,
+    emptyStringIsUndefined: request.query.empty_string_is_undefined,
+  };
+
+  if (reqData.searchAfter === undefined && reqData.searchBefore === undefined) {
+    // simple pagination
+    if (reqData.pageIndex === undefined) {
+      reqData.pageIndex = 0;
+    }
+    reqData.fromIndex = reqData.pageIndex * reqData.pageSize!;
   }
-  return Direction.asc;
+
+  // See: https://github.com/elastic/elasticsearch-js/issues/662
+  // and https://github.com/elastic/endpoint-app-team/issues/221
+  if (
+    reqData.searchBefore !== undefined &&
+    reqData.searchBefore[0] === '' &&
+    reqData.emptyStringIsUndefined
+  ) {
+    reqData.searchBefore[0] = EndpointAppConstants.MAX_LONG_INT;
```

Comment:
why don't these values depend on the sort order?

Reply:
@marshallmain I was confused by that too, but there's a good reason. If you're sorting `asc`, we want them to be at the end, so we'll sort them with the largest possible value. When sorting `desc`, we want them at the beginning... since we're descending, that means they also need to be the largest possible value in this case.

Reply:
ah I misunderstood the purpose of this section, I thought it was setting a default related to the first page of results